docs/architecture/adr-103-proto-versioning.md

```diff
@@ -0,0 +1,140 @@
+v0.37 and v0.38) into `main`, implementing versioning for each major version's
+definitions.
+
+For example, the v0.34 Protobuf definitions would be brought to `main` as `v1`
+(i.e. `proto/tendermint/types/types.proto` would become
+`proto/tendermint/types/v1/types.proto`, and the package would change from
+`tendermint.types` to `tendermint.types.v1`).
+
+Then, the v0.37 Protobuf definitions would be brought to `main`, and only where
+changes were made according to our **sensitive** versioning policy would we
+create new types (e.g. `proto/tendermint/types/params.proto` from v0.37 would
+become `proto/tendermint/types/v2/params.proto` and the package would change
+from `tendermint.types` to `tendermint.types.v2`).
+
+### Minimizing breaking impact
+
+Changing the type URLs of Protobuf message types (e.g. `tendermint.types.Block`
+becoming `tendermint.types.v1.Block`), is considered a breaking change because,
+in some cases, integrators may be serializing structures into Protobuf
+`Any`-typed fields. When doing so, the URL of the type being serialized is also
+embedded in the encoded message.
+
+In order to minimize the impact of this breaking change, it is proposed to keep
+type definitions in the `proto` folder that are _wire-compatible_ with the
+_latest_ versioned types, but that are unversioned (e.g. exposing
+`tendermint.types.Block`, which would be wire-compatible with
+`tendermint.types.v1.Block`).
+
+Internally, however, CometBFT would make use of types generated from the
+versioned Protobuf types only.
+
+This would facilitate a transition period for integrators who depend on the old
+type URL structure, and would allow for a grace period to upgrade to use the
+versioned type definitions.
+
+### Ergonomics of generated code
+
+One of the challenges associated with this change is that usage of the generated
+Go code becomes more difficult for the core team. This is because all generated
+code will also have version suffixes in package paths, meaning that the core
+team needs to know exactly which versions of which messages are relevant to the
+current release when wiring them up.
+
+A simple way of mitigating the impact here is to introduce type aliases for the
+latest versions of generated types, and make use of these aliases internally.
+
+### Rollout strategy
+
+The current recommended strategy to roll out this change is to accumulate all
+incremental changes on a **feature branch**, which can then be targeted to a
+specific release. At present, it is envisaged that CometBFT v0.39 or v0.40 will
+introduce these changes, but this could change based on negotiation with
+stakeholders.
+
+## Consequences
+
+### Positive
+
+- Protobuf definitions that are wire-compatible with all versions of CometBFT
+  will be able to be packaged together
+- The combined Protobuf definitions could be uploaded to the Buf registry for
+  easy consumption by integrators
+- Changes to Protobuf definitions will be more obvious and explicit to
+  integrators
```

Comment:
```suggestion
  integrators. Besides they MUST be documented in the protobuf definitions of newer versions.
```

Reply:
This, I think, should also go into the README in the `proto` folder.